### PR TITLE
feat: scaling factor for pool uptime accumulator to avoid truncation

### PR DESCRIPTION
**Pull request by p0mvn**
## What is the purpose of the change

This is a solution to avoid truncation of the pool reward accumulator to zero when liquidity is large.

It focuses on defining the core of the scaling factor solution ([task ref](https://app.clickup.com/t/86a25h2rg)) and also implementing and testing logic for silently skipping incentive overflows ([task ref](https://app.clickup.com/t/86a25h36d)).

Initial analysis of the approach was done here: https://hackmd.io/o3oqT8VhSPKAiqNl_mlxXQ

To be implemented next:
- [Pool migration](https://app.clickup.com/t/86a25h3et)

We should also analyze whether it makes sense to implement something similar for the spread rewards accumulator by inspecting the new telemetry: https://github.com/osmosis-labs/osmosis/pull/7408

## Testing and Verifying

- Updated unit tests

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/incentives.go [hunk 1]
```diff
@@ -19,6 +19,16 @@ import (
 	"github.com/osmosis-labs/osmosis/v22/x/concentrated-liquidity/types"
 )
 
+// We choose 10^27 to allow sufficient buffer before the accumulator starts getting truncated again.
+// Internally, multiply the number of seconds passed since the last liquidity update by the emission rate per second
```
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:23]:
```suggestion
// Internally, we multiply the number of seconds passed since the last liquidity update by the emission rate per second
```

File: x/concentrated-liquidity/incentives.go [hunk 2]
```diff
@@ -19,6 +19,16 @@ import (
...
+// Then, we scale that value by 10^27 to avoid truncation to zero when dividing by the liquidity in the accumulator.
+// As a result, we do not go for a higher scaling factor to allow for enough room before hitting the maximum integer value of 2^256.
```
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:25]:
```suggestion
// We do not go for a higher scaling factor to allow for enough room before hitting the maximum integer value of 2^256.
```

File: x/concentrated-liquidity/incentives.go [hunk 3]
```diff
@@ -240,11 +250,21 @@ func calcAccruedIncentivesForAccum(ctx sdk.Context, accumUptime time.Duration, l
...
 		// Total amount emitted = time elapsed * emission
-		totalEmittedAmount := timeElapsed.Mul(incentiveRecordBody.EmissionRate)
+		totalEmittedAmount := timeElapsed.MulTruncate(incentiveRecordBody.EmissionRate)
+
+		// We scale up the remaining rewards to avoid truncation to zero
+		// when dividing by the liquidity in the accumulator.
+		scaledTotalEmittedAmount, err := scaleUpTotalEmittedAmount(totalEmittedAmount)
+		if err != nil {
+			ctx.Logger().Info(types.IncentiveOverflowPlaceholderName, "pool_id", poolID, "incentive_id", incentiveRecord.IncentiveId, "time_elapsed", timeElapsed, "emission_rate", incentiveRecordBody.EmissionRate, "error", err.Error())
+			// Silently ignore the truncated incentive record to avoid halting the entire accumulator update
```
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:260]:
```suggestion
			// Silently ignore the truncated incentive record to avoid halting the entire accumulator update.
```

File: x/concentrated-liquidity/keeper_test.go [hunk 4]
```diff
@@ -283,41 +287,52 @@ func (s *KeeperTestSuite) addUptimeGrowthInsideRange(ctx sdk.Context, poolId uin
...
-		newLowerUptimeTrackerValues, err := osmoutils.AddDecCoinArrays(cl.GetUptimeTrackerValues(lowerTickInfo.UptimeTrackers.List), uptimeGrowthToAdd)
+		newLowerUptimeTrackerValues, err := osmoutils.AddDecCoinArrays(cl.GetUptimeTrackerValues(lowerTickInfo.UptimeTrackers.List), growthCopy)
 		s.Require().NoError(err)
 
 		s.initializeTick(ctx, currentTick, lowerTick, lowerTickInfo.LiquidityGross, lowerTickInfo.SpreadRewardGrowthOppositeDirectionOfLastTraversal, wrapUptimeTrackers(newLowerUptimeTrackerValues), true)
 	}
 
 	// In all cases, global uptime accums need to be updated. If currentTick < lowerTick,
 	// nothing more needs to be done.
+	// We do not provide scaled growth here as addToUptimeAccums will scale the growth by the scaling factor.
```
Comment (p0mvn) [inline x/concentrated-liquidity/keeper_test.go:335]:
```suggestion
	// Note that addToUptimeAccums(...) applies the scaling factor to the input before writing to the accumulator.
```

File: x/concentrated-liquidity/incentives.go [hunk 5]
```diff
@@ -19,6 +19,16 @@ import (
 	"github.com/osmosis-labs/osmosis/v22/x/concentrated-liquidity/types"
 )
 
+// We choose 10^27 to allow sufficient buffer before the accumulator starts getting truncated again.
+// Internally, we multiply the number of seconds passed since the last liquidity update by the emission rate per second
+// Then, we scale that value by 10^27 to avoid truncation to zero when dividing by the liquidity in the accumulator.
+// We do not go for a higher scaling factor to allow for enough room before hitting the maximum integer value of 2^256.
+// In the intermediary multiplications.
```
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:26]:
```suggestion
// We do not go for a higher scaling factor to allow for enough room before hitting the maximum integer value of 2^256
// in the intermediary multiplications.
```

File: x/concentrated-liquidity/incentives.go [hunk 6]
```diff
@@ -705,11 +747,22 @@ func (k Keeper) prepareClaimAllIncentivesForPosition(ctx sdk.Context, positionId
...
 
+			// We scale the uptime per-unit of liquidity accumulator up to avoid truncation to zero.
+			// However, once we compute the total for the liquidity entitlement, we must scale it back down.
+			// We always truncate down in the pool's favor.
+			collectedIncentivesForUptime := sdk.NewCoins()
+			for _, incentiveCoin := range collectedIncentivesForUptimeScaled {
+				incentiveCoin.Amount = incentiveCoin.Amount.ToLegacyDec().QuoTruncateMut(perUnitLiqScalingFactor).TruncateInt()
+				if incentiveCoin.Amount.IsPositive() {
+					collectedIncentivesForUptime = append(collectedIncentivesForUptime, incentiveCoin)
+				}
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives.go:763]: Is this to handle the case is which, even after scaling, we still truncate to zero?

File: x/concentrated-liquidity/incentives_test.go [hunk 7]
```diff
@@ -568,14 +571,33 @@ func (s *KeeperTestSuite) TestCalcAccruedIncentivesForAccum() {
...
 
+		"two incentive records, first overflows, second still succeeds": {
+			poolId:               defaultPoolId,
+			accumUptime:          types.SupportedUptimes[0],
+			qualifyingLiquidity:  osmomath.NewDec(100),
+			timeElapsed:          time.Hour,
+			poolIncentiveRecords: []types.IncentiveRecord{withEmissionRate(incentiveRecordOne, oneE60Dec), incentiveRecordOne},
+
+			expectedResult: sdk.DecCoins{
+				// We only expect the first incentive record to qualify
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:590]: We expect both to qualify, just the first is the only that should emit, right? I think the comment might just be wrong here.

File: x/concentrated-liquidity/incentives_test.go [hunk 8]
```diff
@@ -2898,7 +2938,7 @@ func (s *KeeperTestSuite) TestQueryAndClaimAllIncentives() {
 					for _, growthInside := range tc.growthInside {
 						expectedCoins = expectedCoins.Add(sdk.NormalizeCoins(growthInside)...)
 					}
-					s.Require().Equal(expectedCoins, amountClaimed)
+					s.Require().Equal(expectedCoins.String(), amountClaimed.String())
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:2941]: clear indicator there was some debugging going on here 😂

File: x/concentrated-liquidity/incentives_test.go [hunk 9]
```diff
@@ -3557,9 +3603,19 @@ func (s *KeeperTestSuite) TestGetIncentiveRecordSerialized() {
...
+// This test shows that the scaling factor is effective in reducing the likelihood of truncation.
+// However, the scaling factor does not eliminate the possibility of truncation.
 func (s *KeeperTestSuite) TestIncentiveTruncation() {
 	s.SetupTest()
 
+	// We multiply the current tick liqudity by this factor
+	// To bring the current tick liquidity to be at the border line of truncating.
+	// Then, we choose values on each side of the threshold to show that truncation is still possible
+	// but, with the scaling factor, it is less likely to occur due to more room for liquidity to grow.
+	currentTickLiquidityIncreaseFactor := osmomath.BigDecFromDec(cl.PerUnitLiqScalingFactor)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:3617]: Are we able to write out somewhere what exactly these limits are? As in, are we just buying time and if a pool gets 100M liquidity we are back in the same boat, or is the liquidity unlikely (something like 1T for a single pool?)

File: x/concentrated-liquidity/incentives_test.go [hunk 10]
```diff
@@ -3615,8 +3672,35 @@ func (s *KeeperTestSuite) TestIncentiveTruncation() {
 	s.Require().True(incentives.IsZero())
 
 	// Incentives should now be claimed due to lack of truncation
-	s.Ctx = s.Ctx.WithBlockTime(s.Ctx.BlockTime().Add(time.Minute * 51))
+	s.Ctx = s.Ctx.WithBlockTime(s.Ctx.BlockTime().Add(time.Hour * 6))
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:3675]: Why the change of this value?

**Review by czarcas7ic — APPROVED** [review]
Really great work with this, super clear. IMO we should have at minimum 2 approvals on this prior to merge, 3 would be ideal but if we need to move fast I think 2 will suffice.

File: x/concentrated-liquidity/incentives_test.go (hunk 9 above)
Comment (p0mvn) [reply x/concentrated-liquidity/incentives_test.go:3617]: The analysis is in the appendix of the hack md of the PR description.

In a sense, if liquidity growths infinitely, the limit will be hit again.

However, we have alerts in-place. Also, the buffer should be sufficient based on analysis.

Going higher with the scaling factor is also risky because, then, we would be running into overflows

File: x/concentrated-liquidity/incentives.go [hunk 11]
```diff
@@ -240,11 +250,21 @@ func calcAccruedIncentivesForAccum(ctx sdk.Context, accumUptime time.Duration, l
 		}
 
 		// Total amount emitted = time elapsed * emission
-		totalEmittedAmount := timeElapsed.Mul(incentiveRecordBody.EmissionRate)
+		totalEmittedAmount := timeElapsed.MulTruncate(incentiveRecordBody.EmissionRate)
```
Comment (ValarDragon) [inline x/concentrated-liquidity/incentives.go:253]: to future proof this a bit more, thoughts on adding comments in line here for the expected range of each term?

File: x/concentrated-liquidity/incentives.go (hunk 6 above)
Comment (p0mvn) [reply x/concentrated-liquidity/incentives.go:763]: Yeah. It doesn't make sense to collect a bunch of zero coins so decided to simply skip them

File: x/concentrated-liquidity/incentives.go (hunk 11 above)
Comment (p0mvn) [reply x/concentrated-liquidity/incentives.go]: Attempted to address here:
[9e10789](https://github.com/osmosis-labs/osmosis/pull/7409/commits/9e10789efcd423e49320cb9e8389067b28d84cc0)

Let me know what you think

File: x/concentrated-liquidity/incentives_test.go (hunk 10 above)
Comment (p0mvn) [reply x/concentrated-liquidity/incentives_test.go:3675]: The current tick liquidity and scaling factor value were changed so this threshold changes as well.

The goal is to show that with the increased scaling factor, we do not get a truncation within the expected range of increase

File: x/concentrated-liquidity/incentives_test.go (hunk 7 above)
Comment (p0mvn) [reply x/concentrated-liquidity/incentives_test.go]: Right except only the second one qualifies since the first one truncates.

Corrected here: [8a43a0b](https://github.com/osmosis-labs/osmosis/pull/7409/commits/8a43a0b836a5bd7fdfd11eb81a76ecda354b377f)

File: x/concentrated-liquidity/incentives.go [hunk 12]
```diff
@@ -705,11 +784,22 @@ func (k Keeper) prepareClaimAllIncentivesForPosition(ctx sdk.Context, positionId
...
 			if err != nil {
 				return sdk.Coins{}, sdk.Coins{}, err
 			}
 
+			// We scale the uptime per-unit of liquidity accumulator up to avoid truncation to zero.
+			// However, once we compute the total for the liquidity entitlement, we must scale it back down.
+			// We always truncate down in the pool's favor.
+			collectedIncentivesForUptime := sdk.NewCoins()
+			for _, incentiveCoin := range collectedIncentivesForUptimeScaled {
+				incentiveCoin.Amount = incentiveCoin.Amount.ToLegacyDec().QuoTruncateMut(perUnitLiqScalingFactor).TruncateInt()
```
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/incentives.go:797]: To clarify, how does this descaling step account for live pools that will not be migrated? e.g. how do we ensure all unclaimed rewards at launch aren't essentially just zeroed out due to this division by `perUnitLiqScalingFactor` upon claiming?

**Review by AlpinYukseloglu — COMMENTED** [review]
Core scaling logic looks good to me, but had a clarifying question about how pools besides the ones being migrated are being handled

File: x/concentrated-liquidity/incentives.go (hunk 12 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:797]: Discussed offline, resolving since it seems this issue is the focus of #7416 in a separate PR

**Review by AlpinYukseloglu — APPROVED** [review]
LGTM, although I did not think deeply about implications of this change on existing state beyond our brief offline discussion, as it seems that is the focus #7416

File: app/apptesting/test_suite.go [hunk 13]
```diff
@@ -696,3 +696,11 @@ func (s *KeeperTestHelper) SetupVolumeForPools(poolIDs []uint64, volumesForEachP
 func (s *KeeperTestHelper) IncreaseVolumeForPools(poolIDs []uint64, volumesForEachPool []osmomath.Int) {
 	s.SetupVolumeForPools(poolIDs, volumesForEachPool, map[uint64]osmomath.Int{})
 }
+
+// CompareDecCoinsSlice compares two slices of DecCoins
+func (s *KeeperTestHelper) CompareDecCoinsSlice(expected, actual []sdk.DecCoins) {
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply app/apptesting/test_suite.go]: done
